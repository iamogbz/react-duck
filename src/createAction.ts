export function createAction<
    Type extends string = string,
    Payload = unknown,
    State = unknown,
>(
    type: Type,
    prepare = (payload?: Payload): Partial<Action<Type, Payload>> => ({
        payload,
    }),
): ActionCreator<Type, Payload, State> {
    return function createAction(payload?: Payload) {
        const partialAction = prepare(payload);
        return { ...partialAction, type };
    };
}
