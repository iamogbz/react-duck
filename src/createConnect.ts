import { connect } from "./utils/connect";

export function createConnect<S, I, T extends string, P, K>(
    Context?: Context<S, T, P>,
) {
    return function connectContext(
        mapStateToProps?: MapStateToProps<S, I, K>,
        mapDispatchToProps?:
            | MapDispatchToProps<T, P, I>
            | ActionCreatorMapping<T, P, S>,
        mergeProps?: (
            stateProps?: K,
            dispatchProps?: ActionDispatcherMapping<T, P>,
            ownProps?: I,
        ) => I & K & ActionDispatcherMapping<T, P>,
        options?: ConnectOptions<S, T, P, I, K>,
    ): (
        component: ReactComponent<I & K & ActionDispatcherMapping<T, P>>,
    ) => React.FunctionComponent<I> {
        return connect(mapStateToProps, mapDispatchToProps, mergeProps, {
            ...options,
            context: Context,
        });
    };
}
