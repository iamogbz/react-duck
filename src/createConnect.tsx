import React from "react";

export function createConnect<State>(
    Context: React.Context<ContextValue<State>>,
) {
    return function connect<R, S = R, D = R>(
        mapStateToProps?: MapStateToProps<State, R, S>,
        mapDispatchToProps?: MapDispatchToProps<ContextDispatch, R, D>,
    ) {
        return function wrapWithConnect(
            WrappedComponent:
                | React.FunctionComponent<R>
                | React.ComponentClass<R, State>,
        ): React.FunctionComponent<R> {
            const wrappedComponentName =
                WrappedComponent.displayName ||
                WrappedComponent.name ||
                "Component";
            const displayName = `Connect(${wrappedComponentName})`;

            function WrapperComponent(props: R): React.ReactElement {
                const { state, dispatch } = React.useContext(Context);
                const mappedStateProps = React.useMemo(
                    () => mapStateToProps?.(state, props),
                    [state, props],
                );
                const mappedDispatchToProps = React.useMemo(
                    () => mapDispatchToProps?.(dispatch, props),
                    [dispatch, props],
                );
                return (
                    <WrappedComponent
                        {...props}
                        {...mappedStateProps}
                        {...mappedDispatchToProps}
                    />
                );
            }

            WrapperComponent.WrappedComponent = WrappedComponent;
            WrapperComponent.displayName = displayName;
            return WrapperComponent;
        };
    };
}
