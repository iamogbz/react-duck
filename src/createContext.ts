import * as React from "react";
import { SymbolObservable } from "./utils/symbolObservable";
import { setGlobalContext } from "./components/Context";

function createUnimplemented(objectName?: string): (m: string) => () => never {
    const prefix = objectName ? `${objectName}.` : "";
    return function createUnimplemented(methodName) {
        return function unimplemented(): never {
            throw new Error(`Unimplemented method: ${prefix}${methodName}`);
        };
    };
}

export function createContextWithValue<S, T extends string, P>(
    value: Partial<ContextValue<S, T, P>>,
): Context<S, T, P> {
    return React.createContext<ContextValue<S, T, P>>({
        dispatch: (a) => a,
        getState: () => value.state,
        reducer: (s) => s,
        ...value,
    } as ContextValue<S, T, P>);
}

export function createContext<S, T extends string, P>(
    rootReducer: Reducer<S, T, P>,
    preloadedState: S,
    enhancer?: ContextEnhance<S, T, P>,
    displayName?: string,
    global = false,
): Context<S, T, P> {
    const unimplemented = createUnimplemented(`Context(${displayName ?? ""})`);
    const Context = createContextWithValue({
        enhancer,
        reducer: rootReducer,
        state: preloadedState,
        [SymbolObservable]: unimplemented(SymbolObservable.toString()),
    });
    if (displayName) {
        Context.displayName = displayName;
    }
    if (global) {
        setGlobalContext(Context);
    }
    return Context;
}
