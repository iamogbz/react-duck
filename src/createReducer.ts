export function createReducer<State>(
    initialState: State,
    actionHandlers: ActionReducerMapping<State>,
    defaultHandler: Reducer<State> = (s): State => s,
): Reducer<State> {
    return (state = initialState, action?): State => {
        for (const [actionType, reducer] of Object.entries(actionHandlers)) {
            if (actionType === action?.type) {
                return reducer(state, action);
            }
        }
        return defaultHandler(state, action);
    };
}
