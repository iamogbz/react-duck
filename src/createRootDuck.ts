import { combineReducers } from "./utils/combineReducers";

export function createRootDuck<
    D extends Duck<S, N, T, P, R, Q, U, V>[],
    S = any, // eslint-disable-line @typescript-eslint/no-explicit-any
    N extends string = string,
    U extends string = string,
    V extends string = string,
    T extends U | V = U | V,
    P = any, // eslint-disable-line @typescript-eslint/no-explicit-any
    R = any, // eslint-disable-line @typescript-eslint/no-explicit-any
    Q extends string = string
>(...ducks: D): RootDuck<S, N, T, P, R, Q, U, V> {
    const rootDuck = {
        actions: {},
        initialState: {},
        names: new Set(ducks.map((d) => d.name)),
        selectors: {},
    } as RootDuck<S, N, T, P, R, Q, U, V>;
    const reducerMapping = {} as DuckReducerMapping<S, N, T, P>;
    for (const duck of ducks) {
        const duckName = duck.name;
        rootDuck.actions[duckName] = duck.actions;
        rootDuck.initialState[duckName] = duck.initialState;
        rootDuck.selectors[duckName] = duck.selectors;
        reducerMapping[duckName] = duck.reducer;
    }
    rootDuck.reducer = combineReducers(rootDuck.initialState, reducerMapping);
    return rootDuck;
}
