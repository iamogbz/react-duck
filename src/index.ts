export { createAction } from "./createAction";
export { createDuck } from "./createDuck";
export { createReducer } from "./createReducer";
export { createRootReducer } from "./createRootReducer";
