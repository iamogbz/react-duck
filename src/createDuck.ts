import { createReducer } from "./createReducer";
import { createAction } from "./createAction";
import { getEntries } from "./utils/getEntries";
import { combineSelectors } from "./utils/combineSelectors";

function getNS<N extends string, T extends string, V extends string>(
    name: N,
    actionType: T,
): V {
    return `${name}/${actionType}` as V;
}

export function createDuck<
    S,
    P,
    R,
    T extends string,
    U extends string,
    N extends string,
    Q extends string
>({
    name,
    initialState,
    reducers,
    selectors,
    actionMapping,
}: {
    name: N;
    initialState: S;
    reducers: ActionReducerMapping<S, U, P>; // Reducers = { CreatorName: Reducer(State, ActionType, PayloadTypes) }
    selectors?: SelectorMapping<S, R, T, P, Q>;
    actionMapping?: Record<T, U>; // ActionMapping = { MappedActionType: ActionType }
}): Duck<S, N, T, P, R, Q, U> {
    const mappedActionTypes = { ...actionMapping } as Record<T, U>;
    const actions = {} as ActionCreatorMapping<T, P, S, U>;
    for (const [actionType] of getEntries(reducers)) {
        const namespacedActionType = getNS<N, U, T>(name, actionType);
        actions[actionType] = createAction(namespacedActionType);
        mappedActionTypes[namespacedActionType] = actionType;
    }

    const actionReducer = createReducer<S, P, U>(initialState, reducers);
    const isMappedActionType = (a?: Action<string, P>): a is Action<T, P> =>
        a !== undefined &&
        Object.prototype.hasOwnProperty.call(mappedActionTypes, a.type);
    function reducer(state: S, action: Action<T, P> | Action<U, P>): S {
        return actionReducer(state, {
            ...action,
            type: isMappedActionType(action)
                ? mappedActionTypes[action.type]
                : action.type,
        });
    }

    return {
        actions,
        initialState,
        name,
        reducer,
        selectors: combineSelectors(name, selectors),
    };
}
