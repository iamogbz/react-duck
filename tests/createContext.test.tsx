import * as React from "react";
import { act, render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";
import { Provider, applyMiddleware, createContext } from "src";

describe("createContext", () => {
    it("creates context without displayname", () => {
        const Context = createContext((s) => s, {});
        expect(Context.displayName).toBeUndefined();
    });

    it("creates context with displayname", () => {
        const Context = createContext((s) => s, {}, undefined, "TextContext");
        expect(Context.displayName).toBe("TextContext");
    });

    it("has default dispatch and getState", () => {
        const initialState = {};
        const Context = createContext((s) => s, initialState);
        const { result } = renderHook(() => React.useContext(Context));
        expect(result.current.dispatch).not.toThrow();
        expect(result.current.enhancer).toBeUndefined();
        expect(result.current.getState()).toStrictEqual(initialState);
    });

    it("passes context value to consumer", () => {
        const initialState = {};
        const reducer = jest.fn((s) => s);
        const Context = createContext(reducer, initialState);
        const { result } = renderHook(() => React.useContext(Context));
        const contextValue = result.current;
        const spyConsume = jest.fn(() => null);
        render(
            <Provider Context={Context}>
                <Context.Consumer>{spyConsume}</Context.Consumer>
            </Provider>,
        );
        expect(spyConsume).toHaveBeenCalledTimes(1);
        const [consumeValue] = (spyConsume.mock.calls[0] as unknown) as [
            ContextValue,
        ];
        expect(consumeValue.enhancer).toBeUndefined();
        expect(consumeValue.getState()).toStrictEqual(contextValue.state);
        expect(reducer).toHaveBeenCalledTimes(1);
        const action = { type: "ACTION_TYPE" };
        act(() => void consumeValue.dispatch(action));
        expect(reducer).toHaveBeenCalledTimes(2);
        expect(reducer).toHaveBeenLastCalledWith(consumeValue.state, action);
    });

    it("provides and can use empty enhancer from applyMiddleware", () => {
        const initialState = {};
        const Context = createContext(
            (s) => s,
            initialState,
            applyMiddleware(),
        );
        const { result } = renderHook(() => React.useContext(Context));
        const { enhancer, ...value } = result.current;
        expect(enhancer?.(value)).toMatchObject(value);
    });

    it("has expected default context value", () => {
        const Context = createContext((s) => s, {});
        const { result } = renderHook(() => React.useContext(Context));
        expect(result.current).toMatchSnapshot();
        expect(result.current["subscribe"]).toThrow("Unimplemented method");
    });
});
