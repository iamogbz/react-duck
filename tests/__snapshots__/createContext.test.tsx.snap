// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContext has expected default context value 1`] = `
{
  "dispatch": [Function],
  "enhancer": undefined,
  "getState": [Function],
  "reducer": [Function],
  "state": {},
  "subscribe": [Function],
}
`;
